src/syntaxtree.c

```diff
@@ -0,0 +1,1459 @@
+**
+*W  syntaxtree.c
+**
+*/
+#include        <stdarg.h>              /* variable argument list macros   */
+#include        "system.h"              /* Ints, UInts                     */
+
+
+#include        "gasman.h"              /* garbage collector               */
+#include        "objects.h"             /* objects                         */
+#include        "scanner.h"             /* scanner                         */
+
+#include        "gvars.h"               /* global variables                */
+
+#include        "ariths.h"              /* basic arithmetic                */
+#include        "integer.h"
+
+#include        "bool.h"                /* booleans                        */
+
+#include        "gap.h"                 /* error handling, initialisation  */
+
+#include        "calls.h"               /* generic call mechanism          */
+
+#include        "lists.h"               /* generic lists                   */
+
+#include        "records.h"             /* generic records                 */
+#include        "permutat.h"
+#include        "precord.h"             /* plain records                   */
+
+#include        "plist.h"               /* plain lists                     */
+
+#include        "stringobj.h"              /* strings                         */
+
+#include        "code.h"                /* coder                           */
+
+#include        "exprs.h"               /* expressions                     */
+#include        "stats.h"               /* statements                      */
+
+#include        "compiler.h"            /* compiler                        */
+
+#include        "hpc/tls.h"             /* thread-local storage            */
+
+#include        "vars.h"                /* variables                       */
+#include        "compiled.h"
+
+#include <ctype.h>
+
+typedef UInt4 LVar;
+typedef UInt4 HVar;
+typedef UInt GVar;
+
+typedef Obj (*CompileExpr)(Expr expr);
+
+typedef struct {
+  UInt tnum;
+  CompileExpr compiler;
+  const Char *name;
+  UInt arity;
+  const Char *argnames[6];
+} ExprT;
+
+static const ExprT AllExpressions[];
+
+static inline Obj NewSyntaxTreeNode(const char *type)
+{
+    Obj result;
+    Obj typestr;
+
+    C_NEW_STRING_DYN(typestr, type);
+    result = NEW_PREC(1);
+    AssPRec(result, RNamName("type"), typestr);
+
+    return result;
+}
+
+static Obj SyntaxTreeDefaultExpr(Expr expr)
+{
+    Obj result;
+
+    /* TODO: Check that TNUM in range */
+
+    result = NewSyntaxTreeNode(AllExpressions[TNUM_EXPR(expr)].name);
+    return result;
+}
+
+static inline Obj SyntaxTreeExpr(Expr expr)
+{
+    return AllExpressions[TNUM_EXPR(expr)].compiler(expr);
+}
+
+Obj SyntaxTreeUnknownExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("UnknownExpr");;
+    AssPRec(result, RNamName("tnum"), INTOBJ_INT(TNUM_EXPR(expr)));
+    return result;
+}
+
+Obj SyntaxTreeUnknownBool(Expr expr)
+{
+    /* compile the expression and check that the value is boolean          */
```

Comment:
Irrelevant nitpick: Could we perhaps agree to not space pad comments in new code? I think it's a horrible waste of time to go through functions to ensure comments are padded consistently; and if one doesn't do it, it looks just weird. Your TODO comments already follow this, after all (and I'd leave the #include comments alone for now; those are changed by a perl script in the hpc-merge branch anyway)